core/src/main/scala/chisel3/VerificationStatement.scala

```diff
@@ -139,6 +135,36 @@ object assert extends VerifPrintMacrosDoc {
     q"$apply_impl_do($cond, ${getLine(c)}, _root_.scala.None)($sourceInfo)"
   }
 
+  private def emitIfElseFatalIntrinsic(
+    clock:     Clock,
+    predicate: Bool,
+    enable:    Bool,
+    format:    Printable
+  )(
+    implicit sourceInfo: SourceInfo
+  ): Assert = {
+    // Return a list of arguments captured in the printable.
+    def extract_args(pable: Printable): Seq[Bits] = pable match {
+      case Printables(pables) =>
+        val args = pables.map(p => extract_args(p))
+        args.flatten.toSeq
+      case PString(_) | Name(_) | FullName(_) | Percent => List.empty
+      case format: FirrtlFormat => List(format.bits)
+    }
+    val id = Builder.forcedUserModule // It should be safe since we push commands anyway.
+    val data = extract_args(format)
+    val inst = Module(new Chisel_IfElseFatal()(sourceInfo, id, format, data))
```

Comment:
You might be able to use Definition/Instance for this, to limit the number of identical `intmodule`s that gets created in the FIRRTL output. LTL does a similar thing:
- https://github.com/chipsalliance/chisel/blob/main/src/main/scala/chisel3/ltl/LTL.scala#L112-L116

Reply:
We should probably do this, but since the format String is a parameter (rather than an argument), there isn't going to be a lot of benefit. This change is going to be pretty expensive in IR size. @uenoku can you provide some numbers for `.fir` size difference from this change for large designs?

Reply:
Oh, so that's where/how that's done!

With format strings there might not be a huge amount of re-use, especially as long as the strings are likely to contain the source information directly.  Might be good to do anyway, or as a follow-up?

I was briefly concerned about D/I'ing with arguments that need width or reset inference, the idea being if they all instantiate the same the inference behavior might do unexpected things, especially regarding reset.

Not sure what the spec says about these things, but in practice they become operations during parsing where this won't be a problem, just a thought.

Reply:
> I was briefly concerned about D/I'ing with arguments that need width or reset inference, the idea being if they all instantiate the same the inference behavior might do unexpected things, especially regarding reset.

That's a good point, not a problem here but I have had that problem elsewhere. I wonder if we could/should add a way to ask FIRRTL to monomorphize multiply-instantiated things rather than keeping a single instance (I have proposed this as the behavior for future explicit width parameters, but it might be useful to make it specifiable even in that case).

Reply:
I tried using DI but I got an error that `Instanciate` https://github.com/chipsalliance/chisel/blob/ef7de527452df171872349aec0c0493605bea526/core/src/main/scala/chisel3/experimental/hierarchy/Instantiate.scala#L48 requires macros. Is it possible to avoid this kind of error? The FIR text size increase is within 1% in large cores so I think it's fine to use the current approach.

Reply:
You would have to call the [underlying API](https://github.com/chipsalliance/chisel/blob/9448d32685b102f23bd6d97923fc45294635e1d0/core/src/main/scala/chisel3/experimental/hierarchy/Instantiate.scala#L113) of `Instantiate` directly since this is in `core` (the macro is just sugar around calling `Instantiate._impl`). In any case, since this is within 1% and I'm skeptical that D/I will save much here, let's just leave that as future work.